### PR TITLE
Adds Read and Pread to platform.File

### PR DESCRIPTION
**Pull request by codefromthecrypt**
This adds `Read` and `Pread` to `platform.File` backfilling some tests and porting others.

Notably, this no longer supports special casing `io.EOF` when the read buffer is exactly the length of bytes remaining in the file, obviating a follow-up read of length zero. The reason we can't do this internally is `io.EOF` is a special flow control error Go defines, not an actual syscall error. However, it is quite normal to not be able to handle this and the calling code likely would loop around another time for the same reason. Meanwhile the code is a lot less complex as we don't have to test the io.EOF edge case.

Another note is I left stdin alone as I know @achille-roussel and maybe @evacchi are working on non-blocking support there. I didn't want to distract that or cause drift. We can address it later, and meanwhile, I put in a TODO.

See #1417

### CODE REVIEW

**Review by achille-roussel — APPROVED** [review]
Nice work 🙌

**Review by evacchi — APPROVED** [review]
nice!

**Comment by evacchi** [comment]
FYI: I am not working on non-blocking I/O myself yet, because I didn't want to overlap with @achille-roussel (actually we can sync offline)

**Review by mathetake — APPROVED** [review]

File: imports/wasi_snapshot_preview1/fs.go [hunk 1]
```diff
@@ -701,23 +717,21 @@ func fdReadOrPread(mod api.Module, params []uint64, isPread bool) syscall.Errno
...
-	if isPread {
+	var reader func(p []byte) (n int, errno syscall.Errno)
+	if f, ok := fsc.LookupFile(fd); !ok {
+		return syscall.EBADF
+	} else if f.File.AccessMode() == syscall.O_WRONLY {
+		return syscall.EBADF
+	} else if isPread {
 		offset := int64(params[3])
-		reader = sysfs.ReaderAtOffset(r.File.File(), offset)
+		reader = (&preader{f: f.File, offset: offset}).Read
```
[the remark this answers is not in the document]
Comment (codefromthecrypt) [reply imports/wasi_snapshot_preview1/fs.go:731]: we can, just it would might the control flow logic. I can give it a stab after done with the file type, no prob!
